File: src/ReadStorm.Infrastructure/Services/HybridSearchBooksUseCase.cs
using ReadStorm.Application.Abstractions;
using ReadStorm.Domain.Models;

namespace ReadStorm.Infrastructure.Services;

public sealed class HybridSearchBooksUseCase : ISearchBooksUseCase
{
    private readonly RuleBasedSearchBooksUseCase _ruleBased = new();
    private readonly IRuleCatalogUseCase _catalog;

    /// <summary>最多同时并发搜索的书源数量。</summary>
    private const int MaxConcurrentSources = 5;

    /// <summary>全部书源搜索的单源超时。</summary>
    private static readonly TimeSpan PerSourceTimeout = TimeSpan.FromSeconds(12);

    public HybridSearchBooksUseCase(IRuleCatalogUseCase catalog)
    {
        _catalog = catalog;
    }

    public async Task<IReadOnlyList<SearchResult>> ExecuteAsync(
        string keyword,
        int? sourceId = null,
        CancellationToken cancellationToken = default)
    {
        // 指定具体书源时只搜该书源
        if (sourceId is > 0)
        {
            try
            {
                return await _ruleBased.ExecuteAsync(keyword, sourceId, cancellationToken);
            }
            catch
            {
                return [];
            }
        }

        // "全部书源"：并发搜索所有支持搜索的真实书源，汇总去重
        return await SearchAllSourcesAsync(keyword, cancellationToken);
    }

    private async Task<IReadOnlyList<SearchResult>> SearchAllSourcesAsync(
        string keyword, CancellationToken cancellationToken)
    {
        var allRules = await _catalog.GetAllAsync(cancellationToken);
        var searchableRules = allRules
            .Where(r => r.SearchSupported && r.Id > 0)
            .ToList();

        if (searchableRules.Count == 0)
        {
            return [];
        }

        var semaphore = new SemaphoreSlim(MaxConcurrentSources);

        var tasks = searchableRules.Select(async rule =>
        {
            await semaphore.WaitAsync(cancellationToken);
            try
            {
                using var perSourceCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
                perSourceCts.CancelAfter(PerSourceTimeout);
                return await _ruleBased.ExecuteAsync(keyword, rule.Id, perSourceCts.Token);
            }
            catch
            {
                // 单个书源失败不影响整体
                return (IReadOnlyList<SearchResult>)[];
            }
            finally
            {
                semaphore.Release();
            }
        }).ToList();

        var batchResults = await Task.WhenAll(tasks);

        // 按书名+作者去重，保留来源 SourceId
        var merged = batchResults
            .SelectMany(x => x)
            .GroupBy(x => $"{x.Title}|{x.Author}", StringComparer.OrdinalIgnoreCase)
            .Select(g => g.First())
            .Take(100)
            .ToList();

        return merged;
    }
}

File: pureopus/app/src/main/java/com/readstorm/app/infrastructure/services/HybridSearchBooksUseCase.kt
package com.readstorm.app.infrastructure.services

import android.content.Context
import com.readstorm.app.application.abstractions.IRuleCatalogUseCase
import com.readstorm.app.application.abstractions.ISearchBooksUseCase
import com.readstorm.app.domain.models.SearchResult
import kotlinx.coroutines.*
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withPermit

/**
 * Hybrid search: delegates to RuleBasedSearchBooksUseCase for single-source,
 * or searches all searchable sources concurrently when sourceId is null/0.
 */
class HybridSearchBooksUseCase(
    private val context: Context,
    private val catalog: IRuleCatalogUseCase
) : ISearchBooksUseCase {

    private val ruleBased = RuleBasedSearchBooksUseCase(context)

    companion object {
        private const val MAX_CONCURRENT_SOURCES = 5
        private const val PER_SOURCE_TIMEOUT_MS = 12_000L
    }

    override suspend fun execute(keyword: String, sourceId: Int?): List<SearchResult> {
        if (keyword.isBlank()) return emptyList()

        // Single source search
        if (sourceId != null && sourceId > 0) {
            return try {
                ruleBased.execute(keyword, sourceId)
            } catch (e: Exception) {
                AppLogger.log("HybridSearch", "Single source search failed (sourceId=$sourceId): ${e.message}")
                emptyList()
            }
        }

        // All sources search
        return searchAllSources(keyword)
    }

    private suspend fun searchAllSources(keyword: String): List<SearchResult> {
        val allRules = catalog.getAll()
        val searchableRules = allRules.filter { it.searchSupported && it.id > 0 }
        if (searchableRules.isEmpty()) return emptyList()

        val semaphore = Semaphore(MAX_CONCURRENT_SOURCES)

        val results = coroutineScope {
            searchableRules.map { rule ->
                async(Dispatchers.IO) {
                    semaphore.withPermit {
                        try {
                            withTimeout(PER_SOURCE_TIMEOUT_MS) {
                                ruleBased.execute(keyword, rule.id)
                            }
                        } catch (e: Exception) {
                            AppLogger.log("HybridSearch", "Source ${rule.id} failed: ${e.message}")
                            emptyList()
                        }
                    }
                }
            }.awaitAll()
        }

        // Deduplicate by title|author, take top 100
        return results.flatten()
            .distinctBy { "${it.title}|${it.author}".lowercase() }
            .take(100)
    }
}
